File: demos/TodoFilter/src/TodoFilter/Services/TodoFilterService.cs
using TodoFilter.Models;

namespace TodoFilter.Services;

/// <summary>
/// Service for filtering todos by various criteria
/// Demonstrates: SRP, DRY, KISS principles
/// </summary>
public class TodoFilterService
{
    /// <summary>
    /// Filters todos by status (case-insensitive)
    /// </summary>
    /// <param name="todos">List of todos to filter</param>
    /// <param name="status">Status to filter by (e.g., "done", "pending")</param>
    /// <returns>Filtered list of todos</returns>
    public List<Todo> FilterByStatus(List<Todo> todos, string status)
    {
        // Edge case: null or empty inputs
        if (todos == null || string.IsNullOrEmpty(status))
            return new List<Todo>();

        // Case-insensitive matching - users expect "Done" = "done"
        return todos
            .Where(t => t.Status.Equals(status, StringComparison.OrdinalIgnoreCase))
            .ToList();
    }

    /// <summary>
    /// Filters todos created after a specific date
    /// </summary>
    public List<Todo> FilterByDateAfter(List<Todo> todos, DateTime date)
    {
        if (todos == null)
            return new List<Todo>();

        return todos
            .Where(t => t.CreatedAt > date)
            .ToList();
    }

    /// <summary>
    /// Filters todos by title containing search term (case-insensitive)
    /// </summary>
    public List<Todo> FilterByTitleContains(List<Todo> todos, string searchTerm)
    {
        if (todos == null || string.IsNullOrEmpty(searchTerm))
            return new List<Todo>();

        return todos
            .Where(t => t.Title.Contains(searchTerm, StringComparison.OrdinalIgnoreCase))
            .ToList();
    }
}

File: examples/php/src/TodoFilterService.php
<?php

namespace TodoFilter;

/**
 * TodoFilter Service (PHP)
 * Demonstrates: SRP, DRY, KISS principles
 *
 * This is the SAME logic as C# and JavaScript versions,
 * showing that AI-Driven Workflow works across languages
 */
class TodoFilterService
{
    /**
     * Filters todos by status (case-insensitive)
     *
     * @param Todo[] $todos List of todos to filter
     * @param string|null $status Status to filter by
     * @return Todo[] Filtered list
     */
    public function filterByStatus(array $todos, ?string $status): array
    {
        // Edge case: null or empty inputs
        if (empty($todos) || empty($status)) {
            return [];
        }

        // Case-insensitive matching - users expect "Done" = "done"
        return array_values(array_filter(
            $todos,
            fn(Todo $todo) => strcasecmp($todo->status, $status) === 0
        ));
    }

    /**
     * Filters todos created after a specific date
     *
     * @param Todo[] $todos List of todos
     * @param \DateTime|null $date Cutoff date
     * @return Todo[] Filtered list
     */
    public function filterByDateAfter(array $todos, ?\DateTime $date): array
    {
        if (empty($todos) || $date === null) {
            return [];
        }

        return array_values(array_filter(
            $todos,
            fn(Todo $todo) => $todo->createdAt > $date
        ));
    }

    /**
     * Filters todos by title containing search term (case-insensitive)
     *
     * @param Todo[] $todos List of todos
     * @param string|null $searchTerm Term to search for
     * @return Todo[] Filtered list
     */
    public function filterByTitleContains(array $todos, ?string $searchTerm): array
    {
        if (empty($todos) || empty($searchTerm)) {
            return [];
        }

        return array_values(array_filter(
            $todos,
            fn(Todo $todo) => stripos($todo->title, $searchTerm) !== false
        ));
    }
}
